File: runtime/Java/src/org/antlr/v4/runtime/atn/RuleStopState.java
package org.antlr.v4.runtime.atn;

import java.util.ArrayList;
import java.util.List;

/** The last node in the ATN for a rule, unless that rule is the start symbol.
 *  In that case, there is one transition to EOF. Later, we might encode
 *  references to all calls to this rule to compute FOLLOW sets for
 *  error handling.
 */
public class RuleStopState extends ATNState {
	public static final int INITIAL_NUM_TRANSITIONS = 4;

	//public int actionIndex; // for lexer, this is right edge action in rule

	/** Track the transitions emanating from this ATN state. */
	protected List<Transition> transitions =
		new ArrayList<Transition>(INITIAL_NUM_TRANSITIONS);

	@Override
	public int getNumberOfTransitions() { return transitions.size(); }

	@Override
	public void addTransition(Transition e) { transitions.add(e); }

	@Override
	public Transition transition(int i) { return transitions.get(i); }

	@Override
	public void setTransition(int i, Transition e) {
		transitions.set(i, e);
	}
}

File: runtime/JavaScript/src/antlr4/state/RuleStopState.js
/* Copyright (c) 2012-2022 The ANTLR Project Contributors. All rights reserved.
 * Use is of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */
import ATNState from "./ATNState.js";

/**
 * The last node in the ATN for a rule, unless that rule is the start symbol.
 * In that case, there is one transition to EOF. Later, we might encode
 * references to all calls to this rule to compute FOLLOW sets for
 * error handling
 */
export default class RuleStopState extends ATNState {
    constructor() {
        super();
        this.stateType = ATNState.RULE_STOP;
        return this;
    }
}
